### DISCUSSION TITLE
Change the failing system, and probably keep the old one as a mod.

### DISCUSSION THREAD

**Discussion by Doxxz** [ppy/osu#25867] [Ideas]
Sorry for all the edits. I have no idea if discussion edits send notifications to the devs, but if it does I apologize, this is going to be the last one.

### Introduction:

It is apparent that a great part of the playerbase hate HP. Sure, they could just use NF, but as it halves scores, they don't like using it. Some say its better to just remove it, but HP can be useful as an indicator of skill, and there are also many players that love grinding for passes. I believe what they actually hate is the failing system, being forced to stop playing the map because they failed.

### So here is my suggestion: A "DDR style" HP system.

1. If a player's health bar goes to 0%, they "fail" and the play becomes an F, regardless of accuracy, but they can keep going. The game doesn't stop them with a game over screen. One could say this is basically getting Multiplayer's failing system and putting it in Solo.
2. If they continue and eventually complete the map, the F rank play is saved to local leaderboards.
3. If they instead give up before the map ends, they can save the replay from the pause menu.
4. This general concept could be added to every ruleset (for Taiko, maybe we could have it like this: if you complete the map below the pass ratio, you save the F rank).

### Benefits:

- Gives players a reference to beat (their previous F rank scores) when trying to pass maps. A numerical value to gauge their progress.
- Doesn't stop them midway from enjoying the map and song.
- Pass grinders can still retry spam to get their "passes", or rather, non F ranks (we could even add an "Instafail" mod, or setting, that tunnels back to the game over screen).
- Standardizes the failing systems between Multiplayer and Solo.

### About removing NF, and submitting F ranks to global leaderboards:

Some people, like score farmers and completionists, just want their score to be submitted regardless. So keeping NF is necessary to not screw them over.
Another problem is the numerous NF plays already submitted to the leaderboards, they would have to change in some way.

Submitting F ranks is not good either, because if they can overwrite "passed" scores, then sandbagging becomes more viable. It would be like having ranked NF but with no score penalty.

### Conclusion:

Makes gameplay more fun for those who dislike HP. For those that like HP and grinding for passes, it doesn't remove their way of playing the game. Won't ruin score farmers/completionists' way of playing either. Shouldn't have an effect on the servers because it is just an UX change.

**Comment by nana123126** [comment]
'F' should mean 'Failed', and I don't think a failed score should be added to the leaderboard, but continuing the game, I think there can be. The result of failure is failure, otherwise HP is useless. Deleting the HP system is not a good idea, at least a major change is needed, which will make the difficulty level only OD, CS, and AR.

Additionally, NF not only prevents HP from causing failure, but also prevents SD and PF.

So I think what can be added is:

-Do not display the failure screen in case of failure, continue the game and earn an 'F' score after completion
-At any time, press ESC and save a copy of the 'F' score

**Reply by Doxxz** [reply]
Yeah...
Honestly I only added the F rank submission part to please score farmers and to have parity with NF, if we were to remove it.
But the more I think about it the more problematic it all seems.

So it seems that we shouldn't submit F's to online leaderboards, so no score or pp reward for it, and keep NF in for players that just want the score regardless.

This way it already pleases the majority of people that hate HP, because they just hate the game over screen, not really caring about the score.

**Comment by Sleepteiner** [comment]
In my opinion, in a competitive rhythm game, having a pass/fail system that deletes a player's play when failing amounts to basically just a horrible waste of time for the player. "Failing" should still be a thing for people who enjoy going for difficult passes, but I don't think "failing" should have any impact on the player's score. It seems we both agree on all of that.

I don't particularly like the idea of creating a new grade for failed scores though. If someone has 99% accuracy but misaims a single stream near the end, it feels strange to give them a grade lower than D. It would create situations where someone improves an S rank play in both accuracy and miss count, but that improved play is labeled as an F. Personally, I think a better option would be to just mark the score as "failed" on the leaderboard and the results screen without messing with what grade is given to the score.

**Reply by Doxxz** [reply]
"F ranks" are already a thing in the game, but they only show up when submitted to profiles and if you save a failed replay. So I thought it would be easier to just use that.

**Reply by Sleepteiner** [reply]
Ah, true. I forgot about that. But, either way, I personally don't like the idea of using that grade for this specific purpose.

**Comment by bdach** [comment]
> It is apparent that a great part of the playerbase hate HP

It's not apparent to me. Sure there is a vocal *portion* of the community that might not like HP, but I don't know about "apparent". I'd have to see actual polling numbers for this.

From a game design standpoint I don't see the angle here. A fail state is important to gate a beginner player from playing maps way above their skill level. I don't really get the arguments from self-styled "competitive" players as to why failing is bad; if they're really "competitive", then I argue the solution is to "get good".

Almost every other rhythm game I can think of has a definite fail state with definite consequences. Having a fail system with diminished consequences just doesn't make sense to me. At that point I'd rather just remove the mechanic from the game.

Alternative: you can play no fail or difficulty adjust with HP set to 0. With all of the consequences that come with that.

**Reply by Doxxz** [reply]
>but I don't know about "apparent". I'd have to see actual polling numbers for this.

Sorry about this, I only said it out of personal experience from watching people complain about it online, but I have no data.

>From a game design standpoint I don't see the angle here.

I mean, Taiko does something similar and it's doing fine. Seeing how badly they are playing the map, and getting a terrible grade is probably enough for the beginner.

>Almost every other rhythm game I can think of has a definite fail state with definite consequences.

I know DDR, Sound Voltex, Arcaea, Chunithm, Maimai, Wacca, and of course Taiko all have something like this. You need to fill a bar for it to be considered a "pass", or you start on 100% HP and if it goes to 0% you "fail" but you can keep playing the map.

A bit off topic, but if this doesn't get accepted, could we at least have a new feature for taiko that let's you save the score and replay before the map ends? Sometimes I want to quit halfway through, but I still want to save the replay. The problem is that you only fail when the map ends, so if you want the replay you have to wait for the rest of the song to end.

**Reply by Sleepteiner** [reply]
> I don't really get the arguments from self-styled "competitive" players as to why failing is bad; if they're really "competitive", then I argue the solution is to "get good".

The vast majority of tournaments have had NF be required for all players for years now due to the fail system often causing the side with the worse performance in a round to win that round. It is a baffling argument to make to imply that the tournament community is not made up of actual competitive players and that their issues with the fail system are due to them not being skilled enough.

**Reply by bdach** [reply]
@Sleepteiner This discussion is about changing HP in solo, not in realtime multiplayer. I can understand failures not being that strict in realtime multi due to the one shot nature of it, and tournaments are a very narrow use case and what makes sense there does not necessarily make sense for the wider player base. I feel like you are misconstruing my argument, and like I am arguing against a moving target. You mentioned HP being "a waste of time" in your original reply which is only a response that makes sense in the context of solo, so I don't know why my argument is being extended to multi now.

**Reply by Sleepteiner** [reply]
@bdach Realtime multiplayer exacerbates the issues, but the root problem remains the same in both solo and multiplayer. That problem being that the current osu fail system very often rewards a score with a worse performance by deleting scores with a better overall performance because the player did worse on a very small section of the map. For players who mainly enjoy trying to get the best score that they can on a map, this quickly becomes something that wastes a depressing amount of time and effort whether they are in solo play or in a multi.

> You mentioned HP being "a waste of time" in your original reply which is only a response that makes sense in the context of solo

Spending time practicing for a tournament and losing on the tiebreaker despite a better overall performance only because one player on the team misaimed the final stream in the map can be, in my opinion, considered a "waste of time" for many people involved due to the intention of most tournaments being for players to compete for a better performance on the maps played.

**Reply by bdach** [reply]
> That problem being that the current osu fail system very often rewards a score with a worse performance by deleting scores with a better overall performance because the player did worse on a very small section of the map.

That's an aspect of conscious map design. The mapper decides how harsh HP is. If they decide it to be harsh to such a degree, then it is designed to be a challenge. Which is another reason why I'm opposed to lessening or nullifying HP impact.

To use a perhaps somewhat left-of-field comparison, that's like speedrunners asking a game developer to make a skip easier to execute so that they don't need to restart their runs as often if they miss it.

**Reply by Sleepteiner** [reply]
> That's an aspect of conscious map design.

It can be, I agree. That is part of why I'm not advocating for the removal of HP. I'm advocating for failing no longer deleting a score outright. Failing can remain as something that marks a score as "failed" so that players who enjoy going for passes could still do so and the mapper's HP choices on maps are still respected.

> that's like speedrunners asking a game developer to make a skip easier to execute so that they don't need to restart their runs as often if they miss it

I feel like what I'm advocating for is more like a speedrunner asking a game developer to stop the game from kicking the player out to the main menu for making a relatively big mistake in a time trial mode since that run could potentially still be salvaged if the player does well after that mistake. For example, imagine if in the time trial mode in Mario Kart (or other similar games) the player attempts a skip by jumping over a pit, but instead of there just being a time loss for falling in the pit, if the player falls in the pit the game kicks them out to the course select screen regardless of how well they were doing before falling in the pit. If that was the case, I think a lot of players would complain.

**Reply by Doxxz** [reply]
>Failing can remain as something that marks a score as "failed" so that players who enjoy going for passes could still do so and the mapper's HP choices on maps are still respected.

I don't think just a passed marker is a good idea. Failing a map should have consequences, otherwise there is no point. If you don't have the consequence of not submitting the play, then you need to either nerf score or pp.

If you nerf score, then you are basically back to NF, which defeats the purpose I think, and also I'm sure score farmers wouldn't like it. If you nerf pp, then you get into that hyper "sandbagging" scenario I mentioned on the OP. It will make it easier to make plays with less pp, but more score, to overwrite old ones. With how osu! is right now I don't think that is feasible, but if we had plays sorted by pp then it might be a possibility.

On to the other topic, I believe passing a map, in an environment where HP is not unfair (i.e. one mistake leading to an instant fail), can be an exciting challenge. It is an intermediate step between being completely unable to play the map, and "FC'ing" it. It can also be seen as simply a minimum performance requirement to clear the map, set by the mapper.

**Reply by Sleepteiner** [reply]
> Failing a map should have consequences, otherwise there is no point.

The consequences would be that the play is marked as failed and the player wouldn't be able to say that they passed the map and they wouldn't get any pass medals for the play.

> If you nerf score, then you are basically back to NF, which defeats the purpose I think, and also I'm sure score farmers wouldn't like it.

Having NF be essentially auto-applied when failing would significantly increase score farming efficiency. Currently, a score farmer has to either play without NF on a hard map, risking their score being deleted which is very inefficient, or play with NF on a hard map, risking the common situation where they actually ended up passing the map anyway, but losing 50% of the score in the process which is also very inefficient. Being able to play all maps with the potential of getting a passing score and at worst a 50% score penalty if they fail would be great for score farmers. I would be surprised if the average score farmer would be against a change like that if we are solely referring to how it impacts the act of score farming.

For the record, I would still rather not have any score penalty for failing, but a system where failing only applies a score reduction would still be better than the current system for solo play. Although in multi, it would get rid of the concept of recovering from failing which I assume a lot of people would be very against. We could keep recovering from failing as a concept and still just apply the 50% score reduction to the score that actually gets submitted to the solo leaderboard, but that would likely be confusing to players.

> I believe passing a map, in an environment where HP is not unfair (i.e. one mistake leading to an instant fail), can be an exciting challenge.

This wouldn't go away in a system where failed plays are simply marked as failed. Players could still go for exciting passes. The system I'm proposing would keep that aspect of failing for the players who like it while also no longer forcing that aspect of failing on the many players who very much do not consider fighting against the HP system in this game a fun activity.

**Reply by Doxxz** [reply]
>Being able to play all maps with the potential of getting a passing score and at worst a 50% score penalty if they fail would be great for score farmers.

I had not thought about it like that. That sounds good. I somehow forgot that in this proposed system, passed plays wouldn't also get their scores halved. This way we can even remove NF (and screw sandbaggers over yay) and just have this system.

>Although in multi, it would get rid of the concept of recovering from failing which I assume a lot of people would be very against.

Removing recovering makes more sense, but there are probably enough highly controversial changes in this already, so your second idea might be better.

Though now I realized there is a second problem with this multi situation. Tournament players are used to applying NF and not getting their scores halved. What is the play here? Honestly, the only solutions I can think of sound messy.

1. Either they get used to it.
2. Or you keep NF as a legacy tournament mod (similar to scoreV2), that makes plays unranked but with no score penalty.
3. Or you do remove NF, but add a multi lobby setting, like a "Tournament Mode", that does like the above.

Then finally, I think the last question here is what to do about pp in failed plays. Remove it entirely? Keep it all? For reference, in Sound Voltex, when you "crash" a map, your "volforce" gets divided in half. Might be a solution to the "consequences" point from earlier.

**Reply by Sleepteiner** [reply]
> Tournament players are used to applying NF and not getting their scores halved.

There would need to still be some way of turning off failing for tournaments for sure. Also, people have been wanting tournament plays to submit to the global leaderboard for many years now, so the plays shouldn't be unranked either.

> what to do about pp in failed plays.

I think failed plays in this system should just give full pp. In most cases, missing enough to fail is already going to significantly lower the play's pp, so I don't think it needs to be lowered any more artificially.

**Reply by bdach** [reply]
> Having NF be essentially auto-applied when failing would significantly increase score farming efficiency.

I'm sorry but I'm reading that sentence and my brain just shuts down. I hate everything about that sentiment. "farming efficiency" is not something I personally *ever* wanna be promoting in a rhythm game. 

I'm not probably going to reply to this thread anymore because at this point there's a clear chasm in mindset and I don't think I can be reasonably convinced.

**Reply by aiyern** [reply]
> Having NF be essentially auto-applied when failing would significantly increase score farming efficiency.

If I want to score farm, the last thing I'd do is to put NF on. I don't even know why you guys bring this up in the first place.

The currently most efficient way to score farm is those half-an-hour long, couple thousands of objects marathons due to how combo scoring works.

Putting NF on potentially lock you of close to _half a billion score_ with zero difference in difficulty. I personally would put SuddenDeath on since I know I'd have to replay the maps again for a much better score if I don't FC in the first place.

I think there are some arguments that is valid here:  The HP system become more and more annoying to me the harder the maps goes, but bringing up score-farming and completionists, the 2 demographic which probably mostly don't care about HP, sounds really off to me.

**Reply by Doxxz** [reply]
@Sleepteiner 
>Also, people have been wanting tournament plays to submit to the global leaderboard for many years now, so the plays shouldn't be unranked either.

Then I think the only options here are either making tournament players get used to fails halving the score, or we go with the original proposal and keep fails unranked.

@bdach 
>"farming efficiency" is not something I personally ever wanna be promoting in a rhythm game.

Would you still be against the initial proposal? Which is basically just bringing multiplayer's failing system to solo. It's not really promoting score farming, it is just in favor of people that dislike the game over screen.

@aiyern 
>but bringing up score-farming and completionists, the 2 demographic which probably mostly don't care about HP, sounds really off to me.

In my experience, those 2 demographics are the ones that complain about HP the most, for the reasons Sleepteiner brought up above:
>Currently, a score farmer has to either play without NF on a hard map, risking their score being deleted which is very inefficient, or play with NF on a hard map, risking the common situation where they actually ended up passing the map anyway, but losing 50% of the score in the process which is also very inefficient.

**Reply by aiyern** [reply]
> Currently, a score farmer has to either play without NF on a hard map, risking their score being deleted which is very inefficient, or play with NF on a hard map, risking the common situation where they actually ended up passing the map anyway, but losing 50% of the score in the process which is also very inefficient.

There are not that much need to even go for hard maps, assuming 6.5+ SR here, since there already exist a enormous and endless catalogue of easy-to-average difficulties. Playing literally every 5* or below currently exist can probably give you the amount of score comparable to the upper ranks of score farming. There are honestly a lot of ways you can avoid this kind of situation in the first place (some kind of practice diff for the hard parts can help in my experience).

**Reply by Sleepteiner** [reply]
> I'm sorry but I'm reading that sentence and my brain just shuts down. I hate everything about that sentiment. "farming efficiency" is not something I personally ever wanna be promoting in a rhythm game.

I was simply responding to Doxxz's concern. Doxxz was worried that their specific proposal would be disliked by a group of people with a valid playstyle and I responded by saying that I think that group probably wouldn't be too against that proposal since I really don't think the proposal would hurt that specific playstyle. It is very clear that neither Doxxz nor I, despite not agreeing on some of the details, are against the current failing system _because_ it is inefficient for score farming. We are just taking into consideration as many different player groups as we can which is something that people should do when proposing a change to a major mechanic of the game.

**Comment by rodrigopina360** [comment]
I don't really like the idea. A fail is supposed to be a fail, a way to let a user know that they were unable to keep up with the map. I do have a idea that could be good as a middle point here. When you fail give the user the option to resume the play like 5 seconds before the fail, and let them play from that part onwards, until the map ends, and have the score just not submit at all. 
I think it would be cool so that players can replay that section and possibly understand better where they got wrong, without taking the time to open the editor, find the section, and either understand it right from the editor or creating a practice diff of that section. All of this while keeping the exact same penalty that failing would normally bring.

**Reply by Doxxz** [reply]
>When you fail give the user the option to resume the play like 5 seconds before the fail

I think they are going to make a practice mode, which would solve this. 

I don't think if your suggestion is a good idea because players sometimes just want to play the whole map uninterrupted, they wouldn't always want to go back 5 seconds.

**Reply by rodrigopina360** [reply]
Well if you want to play the map uninterupted, either play better and don't fail, or use No Fail, that's probably the best you can do. About practice mode yeah it depends on how it is implemented and how it'll work, both of which afaik haven't been defined yet.

I really just suggested that idea because it's something that doesn't get in the way for users that don't care, is simple enough for users that do care, and in the end doesn't really change anything about the score, keeping the penalty of a fail fully intact.
